File: modules/student-experience/src/main/java/com/campusos/portal/payload/GrievancePayload.java
package com.campusos.portal.payload;

import com.campusos.portal.domain.RequestType;
import java.util.ArrayList;
import java.util.List;

public class GrievancePayload implements RequestPayload {
    public enum Category { ACADEMIC, HOSTEL, EXAM, FEES, OTHER }

    /* ---- frozen contract fields ---- */
    public Category category;
    public String subject;
    public String description;
    public boolean anonymous;

    /* ---- system-derived ---- */
    public Sys sys = new Sys();

    public static class Sys {
        public String routedTo;
    }

    @Override public RequestType type() { return RequestType.GRIEVANCE; }

    @Override public String title() { return subject; }

    @Override public String subtitle() {
        return category.name().charAt(0) + category.name().substring(1).toLowerCase()
                + (sys.routedTo == null ? "" : " · " + sys.routedTo)
                + (anonymous ? " · anonymous" : "");
    }

    @Override public List<Artifact> artifacts() { return new ArrayList<>(); }

    /** The routed desk, not the generic matrix actor — "Hostel Warden Office", not "Class Advisor". */
    @Override public String handledBy() {
        return sys.routedTo == null || sys.routedTo.isBlank() ? null : sys.routedTo;
    }

    @Override public void validate() {
        if (category == null) throw new IllegalArgumentException("category is required");
        if (subject == null || subject.isBlank()) throw new IllegalArgumentException("subject is required");
        if (description == null || description.isBlank()) throw new IllegalArgumentException("description is required");
    }
}

File: modules/student-experience-js/src/payload/GrievancePayload.js
import { RequestType, GrievanceCategory } from '../domain/enums.js';
import { IllegalArgumentException } from './RequestPayload.js';
import { DisplayLabels } from '../view/DisplayLabels.js';

export class GrievancePayload {
  constructor() {
    this.category = null;
    this.subject = null;
    this.description = null;
    this.anonymous = false;
    this.sys = { routedTo: null };
  }

  type() { return RequestType.GRIEVANCE; }

  title() { return this.subject; }

  subtitle() {
    return `${DisplayLabels.category(this.category)} · ${DisplayLabels.desk(this.category)}`
      + (this.anonymous ? ' · anonymous' : '');
  }

  artifacts() { return []; }

  /**
   * The desk, not the generic matrix actor — "Hostel Warden Office", never "Class Advisor".
   * Derived from the category at render time; nothing in the engine maintains sys.routedTo.
   */
  handledBy() { return DisplayLabels.desk(this.category); }

  validate() {
    if (!this.category || !Object.values(GrievanceCategory).includes(this.category)) {
      throw new IllegalArgumentException('category is required');
    }
    if (!this.subject || !this.subject.trim()) throw new IllegalArgumentException('subject is required');
    if (!this.description || !this.description.trim()) throw new IllegalArgumentException('description is required');
  }

  static fromJSON(o) { return Object.assign(new GrievancePayload(), o, { sys: { ...new GrievancePayload().sys, ...o.sys } }); }
}
